### ISSUE TITLE
Page allocator should have option to be consistent with EDKII

### ISSUE THREAD

**Issue by cfernald** [OpenDevicePartnership/patina#191]
### Is there an existing issue for this?

- [x] I have searched existing issues

### Current Behavior

The current page allocator default is to do a pure top-down allocation. This was done to avoid allocating the <1MB memory on x64 which is needed for MP support. However, this allocation scheme is different from the C code allocator which does a scope bottom-up allocation. This has exposed a number of bugs across UEFI and Windows, but before releasing it would be good to go for consistency to avoid incompatibility with 3rd party software and drivers.

### Build Environment

```markdown
- OS(s):
- Tool Chain(s):
- Targets Impacted:
```

### Version Information

```text

```

### Urgency

Low

### Are you going to fix this?

Someone else needs to fix it

### Do you need maintainer feedback?

None

**Comment by makubacki** [comment]
@cfernald, do you still plan to work on this?

**Comment by cfernald** [comment]
Yes, I will take a look at this soon.

**Comment by makubacki** [comment]
@cfernald, bumping as a reminder.

**Comment by makubacki** [comment]
@cfernald, do you think we can include the "switch" to modify the allocation direction in this issue's work? I assume that's going to be done anyway to preserve the ability to test top-down.

**Comment by cfernald** [comment]
@makubacki, I completely agree. I'm thinking a `.with_allocation_pattern` or similar routine for the core. We could also have a feature flag or always default to this behavior in the debug profile, but I don't think this would warrant a feature flag and using the profile may be confusing and prevents the platform from being able to ship with this configuration if they prefer. Obviously this does require the platform to intentionally enable this though which might mean many platforms may never do so.

perhaps it would be good for us to lump features indented to expose bugs together behind a single configuration call such as this, heap validations, certain more robust memory protections, etc.

**Comment by cfernald** [comment]
As discussed offline, the current plan would be to by default use the best configuration for testing and provide a config function from the core to choose an alternative. So, we will continue to use the top-down allocation by default but allow the platform to use the EDKII like allocation pattern when it is appropriate for that platform.

**Comment by makubacki** [comment]
And in the documentation mention that top-down is the recommended configuration. A platform should only use bottom up if an actual issue has been encountered with top-down and they should track the switch to bottom-up with this config option as a workaround that can be undone after the blocking issue is resolved.

**Comment by cfernald** [comment]
@makubacki, as a heads up on this I decided to take a slightly different strategy to simplify the implementation but would like to get your thoughts. Instead of having a exposed concept of top-down, bottom-up. Instead I would rather always go top-down, and expose a configuration to "prioritize 32-bit memory." This configuration would cause us to take any any_pages allocation request and will first try a max_addres(0xFFFF_FFFF). This should give very similar allocation to EDKII and is more explicit about what it is doing instead of trying to match EDKII explicitly. What do you think?

**Comment by makubacki** [comment]
That sounds good to me. The consistency within Patina is nice and, as you mentioned, it better describes the actual need.
